File: caldav_account.js
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

// The CalDAV account, again, for a browser: what is on it, and which calendar
// is the default. `caldav_account.py` is the other half of this pair, and the
// two are held together by the fixtures they are both asserted against --
// `fixtures/caldav-home-*.xml` and its `.expected.json` companions. Change
// what one of them decides and the other's tests fail, which is the point.
//
// The XML is read by the small parser at the bottom of this file rather than by
// DOMParser, for two reasons. It runs identically under `node --test` and in the
// add-on, so the tests exercise the code that ships; and a server's XML never
// reaches a DOM API at all, which is a stronger version of the rule that it must
// never be assigned into the live document.

export const DAV = "DAV:";
export const CALDAV = "urn:ietf:params:xml:ns:caldav";

// What a calendar with no name is called, matching UNNAMED in caldav_account.py.
export const UNNAMED = "(unnamed)";

// What is in the account: every child collection, what kind it is, and what it
// is called. Byte for byte the body caldav_account.py sends.
export const LISTING = `<?xml version="1.0" encoding="utf-8"?>
<D:propfind xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">
  <D:prop>
    <D:resourcetype/>
    <D:displayname/>
    <D:current-user-principal/>
    <C:schedule-default-calendar-URL/>
  </D:prop>
</D:propfind>
`;

// Which calendar the account treats as its default, asked of the principal when
// the home did not say.
export const DEFAULT = `<?xml version="1.0" encoding="utf-8"?>
<D:propfind xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">
  <D:prop>
    <C:schedule-default-calendar-URL/>
  </D:prop>
</D:propfind>
`;

/** The MKCALENDAR body for a calendar of this name, as caldav_make_calendar.py sends it. */
export function makeBody(name) {
  return `<?xml version="1.0" encoding="utf-8"?>
<C:mkcalendar xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">
  <D:set><D:prop><D:displayname>${escapeXml(name)}</D:displayname></D:prop></D:set>
</C:mkcalendar>
`;
}

/** A href as a bare path, however the server chose to write it. */
export function pathOf(href) {
  const raw = (href || "").trim();
  if (!raw) return "";
  let path = raw;
  if (/^[a-z][a-z0-9+.-]*:\/\//i.test(raw)) {
    try {
      path = new URL(raw).pathname;
    } catch {
      path = raw;
    }
  }
  return path.replace(/\/+$/, "");
}

/**
 * Every calendar in a multistatus reply, and which one the server calls default.
 *
 * The second half is null far more often than the specification suggests: it is
 * what the server *said*, and Stalwart says nothing. Pass both to defaultAmong()
 * rather than reading the null as "there isn't one".
 */
export function calendarsIn(xml) {
  const root = parseXml(xml);
  const calendars = [];
  let advertised = null;
  let principal = null;

  for (const response of findAll(root, DAV, "response")) {
    const href = textOf(find(response, DAV, "href"));
    const kind = find(response, DAV, "resourcetype");
    if (principal === null) {
      principal = textOf(find(find(response, DAV, "current-user-principal"), DAV, "href")) || null;
    }
    if (advertised === null) {
      const said = find(response, CALDAV, "schedule-default-calendar-URL");
      advertised = textOf(find(said, DAV, "href")) || null;
    }
    if (!href || !kind) continue;
    // A calendar, and specifically not the scheduling inbox or outbox, which
    // are collections in the same place and would break the account if they
    // went. Nothing here deletes anything, but a listing that calls the inbox a
    // calendar is a listing that invites somebody else's code to.
    if (!find(kind, CALDAV, "calendar")) continue;
    if (find(kind, CALDAV, "schedule-inbox")) continue;
    if (find(kind, CALDAV, "schedule-outbox")) continue;
    const name = textOf(find(response, DAV, "displayname"));
    calendars.push({ href, path: pathOf(href), name: name || UNNAMED });
  }

  return {
    calendars,
    advertised: advertised ? pathOf(advertised) : null,
    principal: principal ? pathOf(principal) : null,
  };
}

/** The one href a schedule-default-calendar-URL reply carries, or null. */
export function defaultIn(xml) {
  const said = find(parseXml(xml), CALDAV, "schedule-default-calendar-URL");
  const href = textOf(find(said, DAV, "href"));
  return href ? pathOf(href) : null;
}

/**
 * Which calendar is the default, and whether the server actually said so.
 *
 * Thundermail's Stalwart advertises schedule-default-calendar-URL nowhere, so on
 * the server these tools were written for the answer always comes from the
 * address ending in /default. That is a guess, and callers have to be able to
 * tell the two apart: naming the wrong calendar as your default sends somebody
 * to test against the one calendar they cannot delete afterwards.
 *
 * The name is deliberately not consulted. A calendar can be called anything,
 * "Default" included, and the display name is the last thing that tells you
 * which calendar the account schedules into.
 */
export function defaultAmong(calendars, advertised) {
  if (advertised) return { path: advertised, said: true };
  for (const calendar of calendars) {
    if (calendar.path.split("/").pop().toLowerCase() === "default") {
      return { path: calendar.path, said: false };
    }
  }
  return { path: null, said: false };
}

/**
 * The last part of a calendar's address, worked out from its name.
 *
 * Names have spaces, accents and punctuation in them and addresses should not,
 * so this keeps the letters and digits and joins the rest with hyphens --
 * character for character what address_for() does in caldav_make_calendar.py.
 */
export function addressFor(name) {
  const slug = (name || "").replace(/[^A-Za-z0-9]+/g, "-").replace(/^-+|-+$/g, "").toLowerCase();
  return slug || "calendar";
}

// Where Thundermail keeps an account's calendars. Read off a real account on
// 2026-09-01 -- nemo@thundermail.com's calendars are at
// https://mail.thundermail.com/dav/cal/nemo%40thundermail.com/ -- so it is one
// observation rather than a documented rule, which is why it only ever fills a
// field in that you can then change.
const THUNDERMAIL = { domain: "thundermail.com", host: "mail.thundermail.com", dav: "/dav/cal/" };

/**
 * The address of that account's calendars, where it can be worked out.
 *
 * Only Thundermail, and only because it has been looked at. Every other
 * provider puts its calendars somewhere of its own choosing, and guessing at a
 * path that turns out to be wrong wastes more of your time than an empty field
 * does: a 404 from the wrong address looks exactly like an account with no
 * calendars in it.
 */
export function homeFor(user) {
  const address = (user || "").trim();
  const at = address.lastIndexOf("@");
  if (at < 1) return null;
  if (address.slice(at + 1).toLowerCase() !== THUNDERMAIL.domain) return null;
  // The address is a path segment here, so its @ has to be written %40.
  return `https://${THUNDERMAIL.host}${THUNDERMAIL.dav}${encodeURIComponent(address)}/`;
}

/** Whether that host is a server here rather than one out on the internet. */
export function onThisMachine(host) {
  const name = (host || "").replace(/^\[|\]$/g, "").toLowerCase().replace(/\.$/, "");
  if (name === "localhost" || /\.(localhost|local|internal)$/.test(name)) return true;
  if (name === "::1" || name === "0:0:0:0:0:0:0:1") return true;
  const parts = name.split(".");
  if (parts.length === 4 && parts.every((p) => /^\d{1,3}$/.test(p) && Number(p) < 256)) {
    const [a, b] = parts.map(Number);
    return a === 127 || a === 10 || (a === 192 && b === 168) || (a === 172 && b >= 16 && b <= 31);
  }
  return false;
}

/**
 * Whether this address may be asked for a calendar list, and why not.
 *
 * Plain HTTP to somebody's mail server puts an app password on the wire, so it
 * is allowed only where the wire is this machine. That is the same line
 * caldav_import_ics.py draws for --unscrubbed, and it is drawn here in code for
 * the same reason: a rule written only in the documentation is a rule that gets
 * read after the request has gone.
 */
export function targetProblem(address) {
  let url;
  try {
    url = new URL(address);
  } catch {
    return "That does not look like a web address. It should start with https:// and end with a /";
  }
  if (url.protocol === "https:") return null;
  if (url.protocol !== "http:") return `${url.protocol} is not an address a calendar lives at.`;
  if (onThisMachine(url.hostname)) return null;
  return (
    "That address is plain http, which would send your app password across the network " +
    "unencrypted. Use https, unless the server is on this machine."
  );
}

/** The Authorization header for a username and an app password, non-ASCII included. */
export function basicAuth(user, password) {
  const bytes = new TextEncoder().encode(`${user}:${password}`);
  let binary = "";
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return `Basic ${btoa(binary)}`;
}

export function escapeXml(text) {
  return String(text)
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&apos;");
}

// --------------------------------------------------------------------------
// Reading the XML.
//
// Enough of a parser for a multistatus reply and no more: elements, text,
// attributes read only for their xmlns declarations, and the five entities a
// display name can contain. Namespaces are resolved rather than assumed,
// because the prefix is the server's choice -- D:, A:, d:, or none at all --
// and a parser that keys on "D:href" works until the day a server changes it.

const NAME = "[^\\s/>=]+";
const TAG = new RegExp(`<(/?)(${NAME})((?:[^>"']|"[^"]*"|'[^']*')*?)(/?)>`, "g");
const ATTRIBUTE = new RegExp(`(${NAME})\\s*=\\s*("[^"]*"|'[^']*')`, "g");

/** The document as a tree of {ns, name, children, text}. */
export function parseXml(text) {
  const source = String(text)
    .replace(/<\?[\s\S]*?\?>/g, "")
    .replace(/<!--[\s\S]*?-->/g, "")
    .replace(/<!DOCTYPE[^>]*>/gi, "");

  const root = { ns: null, name: "#document", children: [], text: "" };
  const stack = [{ node: root, namespaces: {} }];
  let last = 0;
  TAG.lastIndex = 0;

  for (let tag = TAG.exec(source); tag; tag = TAG.exec(source)) {
    const [whole, closing, qualified, attributes, selfClosing] = tag;
    const between = source.slice(last, tag.index);
    last = tag.index + whole.length;
    if (between.trim()) stack[stack.length - 1].node.text += decode(between);

    if (closing) {
      if (stack.length > 1) stack.pop();
      continue;
    }

    const namespaces = { ...stack[stack.length - 1].namespaces };
    ATTRIBUTE.lastIndex = 0;
    for (let attr = ATTRIBUTE.exec(attributes); attr; attr = ATTRIBUTE.exec(attributes)) {
      const value = decode(attr[2].slice(1, -1));
      if (attr[1] === "xmlns") namespaces[""] = value;
      else if (attr[1].startsWith("xmlns:")) namespaces[attr[1].slice(6)] = value;
    }

    const colon = qualified.indexOf(":");
    const prefix = colon === -1 ? "" : qualified.slice(0, colon);
    const node = {
      ns: namespaces[prefix] || null,
      name: colon === -1 ? qualified : qualified.slice(colon + 1),
      children: [],
      text: "",
    };
    stack[stack.length - 1].node.children.push(node);
    if (!selfClosing) stack.push({ node, namespaces });
  }

  return root;
}

/** Element names are compared case-insensitively: schedule-default-calendar-URL. */
function matches(node, ns, name) {
  return node.ns === ns && node.name.toLowerCase() === name.toLowerCase();
}

/**
 * Every descendant with this name, at any depth, in document order.
 *
 * Depth rather than a fixed path because the shape above a <response> is the
 * server's business: parseXml() hands back the document, the multistatus sits
 * under it, and a server is free to wrap things differently again.
 */
export function findAll(node, ns, name, found = []) {
  if (!node) return found;
  for (const child of node.children) {
    if (matches(child, ns, name)) found.push(child);
    else findAll(child, ns, name, found);
  }
  return found;
}

/** The first descendant with this name, at any depth, or null. */
export function find(node, ns, name) {
  if (!node) return null;
  for (const child of node.children) {
    if (matches(child, ns, name)) return child;
    const deeper = find(child, ns, name);
    if (deeper) return deeper;
  }
  return null;
}

export function textOf(node) {
  return node ? node.text.trim() : "";
}

function decode(text) {
  return text.replace(/&(#x?[0-9a-f]+|[a-z]+);/gi, (whole, body) => {
    if (body[0] === "#") {
      const code = body[1] === "x" || body[1] === "X"
        ? parseInt(body.slice(2), 16)
        : parseInt(body.slice(1), 10);
      return Number.isFinite(code) ? String.fromCodePoint(code) : whole;
    }
    const named = { amp: "&", lt: "<", gt: ">", quot: '"', apos: "'" }[body.toLowerCase()];
    return named === undefined ? whole : named;
  });
}

File: caldav_account.py
#!/usr/bin/env python3
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

"""What every CalDAV tool here needs before it can do its own job: a connection
to the server, and the list of calendars on the account.

Neither of those is about deleting anything, but both of them used to live in a
tool that does. That was fine while every tool in the set changed something --
each one had to sign in, list, and only then delete, import or create. A tool
that only looks is the case that made it wrong: caldav_list_calendars.py would
have had to import two deleting tools to ask the server a question.

So the transport is Connection, the listing is Account, and the tools that
change something add their own verb on top:

    Connection      one server, one collection, one connection kept open
      Account       the calendars on it, and which one is the default
        Deleter     ... and DELETE, in caldav_delete_calendars.py
        Maker       ... and MKCALENDAR, in caldav_make_calendar.py
      Calendar      one calendar's entries, in caldav_delete_events.py
        Importer    ... and PUT, in caldav_import_ics.py

The credentials that open one of these come from caldav_asking.py, which is the
other half of what the tools share.
"""

from __future__ import annotations

import base64
import http.client
import xml.etree.ElementTree as ET
from urllib.parse import urlsplit

DAV = "DAV:"
CALDAV = "urn:ietf:params:xml:ns:caldav"

# What is in the account: every child collection, what kind it is, and what it
# is called. The kind matters -- an account's calendar home also holds the
# scheduling inbox and outbox, which are collections but are not calendars and
# must not be deleted.
LISTING = """<?xml version="1.0" encoding="utf-8"?>
<D:propfind xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">
  <D:prop>
    <D:resourcetype/>
    <D:displayname/>
    <D:current-user-principal/>
    <C:schedule-default-calendar-URL/>
  </D:prop>
</D:propfind>
"""

# Which calendar the account treats as its default. Servers put this in
# different places, so it gets asked for wherever it might be.
DEFAULT = """<?xml version="1.0" encoding="utf-8"?>
<D:propfind xmlns:D="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">
  <D:prop>
    <C:schedule-default-calendar-URL/>
  </D:prop>
</D:propfind>
"""

# What a calendar with no name is called, so it still has something to print and
# something to match --keep against.
UNNAMED = "(unnamed)"


def _path_of(href: str) -> str:
    """A href as a bare path, however the server chose to write it.

    Three shapes arrive: a path, a full URL, and a network-path reference --
    //host/path, which is a host without a scheme. urlsplit reduces all three to
    the path, and that is load-bearing rather than incidental: a host named in a
    *reply* is not one to act on, and the JavaScript twin had a hole here
    exactly because it only recognised the second shape.
    """
    return (urlsplit(href).path or href).rstrip("/")


class Connection:
    """One collection on a CalDAV server, over a connection kept open."""

    def __init__(self, url: str, user: str, password: str, timeout: float = 60):
        parts = urlsplit(url)
        if parts.scheme not in ("http", "https"):
            raise ValueError(f"{url} is not an http or https address")
        if not parts.path:
            raise ValueError(f"{url} has no path, so it cannot be a calendar")
        self.secure = parts.scheme == "https"
        self.host = parts.hostname or ""
        self.port = parts.port
        self.timeout = timeout
        # A calendar is a collection, so its address ends in a slash. Servers
        # differ on whether they forgive a missing one; none mind an extra.
        self.path = parts.path if parts.path.endswith("/") else parts.path + "/"
        credentials = f"{user}:{password}".encode()
        self.authorization = "Basic " + base64.b64encode(credentials).decode("ascii")
        self.connection: http.client.HTTPConnection | None = None

    def address(self) -> str:
        """The calendar's address as the tool understands it, to print back at you.

        Assembled from the parts rather than echoed, because what a request goes
        to is the path with its trailing slash rather than whatever was typed.
        Which means the scheme and the port are assembled too, and neither is a
        detail to lose: a local server on some port is exactly where a
        reproduction that must not touch production goes.
        """
        port = f":{self.port}" if self.port else ""
        return f"{'https' if self.secure else 'http'}://{self.host}{port}{self.path}"

    def _connect(self) -> http.client.HTTPConnection:
        if self.connection is None:
            opener = http.client.HTTPSConnection if self.secure else http.client.HTTPConnection
            self.connection = opener(self.host, self.port, timeout=self.timeout)
        return self.connection

    def close(self) -> None:
        if self.connection is not None:
            self.connection.close()
            self.connection = None

    def request(
        self, method: str, path: str, body: str | None = None, headers: dict | None = None
    ) -> tuple[int, bytes]:
        """Make one request, reconnecting once if the server dropped us.

        Hundreds of deletions go down a single connection and servers close idle
        ones, so a dropped connection is routine rather than a failure.
        """
        head = {"Authorization": self.authorization, "User-Agent": "desktop-support-tools"}
        if body is not None:
            head["Content-Type"] = "application/xml; charset=utf-8"
        head.update(headers or {})
        payload = body.encode("utf-8") if body is not None else None

        for attempt in (1, 2):
            try:
                connection = self._connect()
                connection.request(method, path, body=payload, headers=head)
                response = connection.getresponse()
                return response.status, response.read()
            except (http.client.HTTPException, OSError):
                self.close()
                if attempt == 2:
                    raise
        raise AssertionError("unreachable")

    def _multistatus(self, method: str, body: str, depth: str) -> tuple[ET.Element, bool]:
        """Make a request that returns a list, and say whether it was cut short.

        A server that will not hand back everything says so rather than quietly
        stopping, and taking a short reply for a short calendar is how you end
        up deleting half of what you meant to.
        """
        status, raw = self.request(method, self.path, body, {"Depth": depth})
        if status == 401:
            raise SystemExit(
                "The server would not accept that username and password.\n"
                "CalDAV here needs an app password; an OAuth2 account password will not do.\n"
                "Some servers also want the bare username rather than the full address."
            )
        if status == 404:
            raise SystemExit(f"There is no calendar at {self.path} -- check the Location field.")
        if status != 207:
            raise SystemExit(f"Asking the server about the calendar failed: HTTP {status}")
        try:
            root = ET.fromstring(raw)
        except ET.ParseError as error:
            raise SystemExit(f"The server's reply was not readable XML: {error}") from None

        cut_short = root.find(f".//{{{DAV}}}number-of-matches-within-limits") is not None
        return root, cut_short


class Account(Connection):
    """An account's calendars, reached over the connection Connection opens."""

    #: Where the last listing said the account's principal is, as a path.
    principal: str | None = None

    def calendars(self) -> tuple[list[tuple[str, str]], str | None]:
        """Every calendar as its address and name, and which one the server calls default.

        The second half is None far more often than the specification suggests:
        it is what the server *said*, and Stalwart says nothing. Pass both to
        default_among() rather than reading the None as "there isn't one".
        """
        root, _ = self._multistatus("PROPFIND", LISTING, "1")

        default = None
        found = []
        principal = None
        for response in root.findall(f"{{{DAV}}}response"):
            href = (response.findtext(f"{{{DAV}}}href") or "").strip()
            kind = response.find(f".//{{{DAV}}}resourcetype")
            if principal is None:
                principal = response.findtext(
                    f".//{{{DAV}}}current-user-principal/{{{DAV}}}href"
                )
            if default is None:
                default = response.findtext(
                    f".//{{{CALDAV}}}schedule-default-calendar-URL/{{{DAV}}}href"
                )
            if not href or kind is None:
                continue
            # A calendar, and specifically not the scheduling inbox or outbox,
            # which are collections in the same place and would break the
            # account if they went.
            if kind.find(f"{{{CALDAV}}}calendar") is None:
                continue
            if kind.find(f"{{{CALDAV}}}schedule-inbox") is not None:
                continue
            if kind.find(f"{{{CALDAV}}}schedule-outbox") is not None:
                continue
            name = (response.findtext(f".//{{{DAV}}}displayname") or "").strip()
            found.append((href, name or UNNAMED))

        # Kept rather than returned, to leave calendars() a pair: every caller
        # wants the calendars and the default, and only the tests and the
        # add-on's twin care where the principal was.
        self.principal = _path_of(principal) if principal else None

        if default is None and principal:
            default = self._default_from(principal)
        return found, _path_of(default) if default else None

    def _default_from(self, principal: str) -> str | None:
        """Ask the account's principal which calendar is the default one."""
        here = self.path
        try:
            self.path = _path_of(principal) + "/"
            root, _ = self._multistatus("PROPFIND", DEFAULT, "0")
        except SystemExit:
            return None
        finally:
            self.path = here
        return root.findtext(f".//{{{CALDAV}}}schedule-default-calendar-URL/{{{DAV}}}href")


def default_among(
    calendars: list[tuple[str, str]], advertised: str | None
) -> tuple[str | None, bool]:
    """Which calendar is the default, and whether the server actually said so.

    Thundermail's Stalwart advertises schedule-default-calendar-URL nowhere --
    not on the home, not on the principal -- so on the server these tools were
    written for the answer always comes from the address ending in /default.
    That is a guess, and every caller has to be able to tell the callers apart:
    guessing wrong means keeping a calendar you meant to delete, which is
    harmless, or naming the wrong calendar as your default, which is not.

    The name is deliberately not consulted. A calendar can be called anything,
    including "Default", and the one thing a display name never tells you is
    which calendar the account will fall back to.
    """
    if advertised:
        return advertised, True
    for href, _ in calendars:
        path = _path_of(href)
        if path.rsplit("/", 1)[-1].casefold() == "default":
            return path, False
    return None, False
